File: frontend/src/api/reports.ts
import { apiFetch } from "./client";

export interface CoverageByArea {
  area_id: string;
  area_name: string;
  days_covered: number;
  days_uncovered: number;
  coverage_pct: number;
  gaps: { date: string; day_name: string }[];
}

export interface CoverageResponse {
  period_label: string;
  overall_coverage_pct: number;
  total_gaps: number;
  most_critical_area: string | null;
  weakest_day: string | null;
  by_area: CoverageByArea[];
}

export interface WorkloadEntry {
  doctor_id: string;
  name: string;
  rank: string | null;
  sex: string | null;
  department: string | null;
  emergencia: number;
  pista: number;
  disponible: number;
  total: number;
  details: { date: string; area: string }[];
}

export interface WorkloadResponse {
  period_label: string;
  total_services: number;
  active_doctors: number;
  avg_per_doctor: number;
  most_load: { name: string; total: number } | null;
  least_load: { name: string; total: number } | null;
  entries: WorkloadEntry[];
}

export interface DossierResponse {
  doctor_id: string;
  name: string;
  rank: string | null;
  sex: string | null;
  department: string | null;
  areas: string[];
  period_label: string;
  total_services: number;
  services_by_area: Record<string, number>;
  avg_weekly: number;
  services: { date: string; day_name: string; area: string; source: string }[];
  missions: { mission: string; role: string; status: string }[];
  restrictions: { type: string; date: string | null; reason: string }[];
  availability: string[];
}

interface CoverageParams {
  year_start: number; month_start: number;
  year_end: number; month_end: number;
  area?: string; rank_id?: string; sex?: string; department_id?: string;
}

interface WorkloadParams {
  year: number; month: number;
  area?: string; rank_id?: string; sex?: string; department_id?: string;
  group_by?: string; order_by?: string;
}

export const reportsApi = {
  getCoverage(params: CoverageParams): Promise<CoverageResponse> {
    const q = new URLSearchParams();
    Object.entries(params).forEach(([k, v]) => { if (v !== undefined) q.set(k, String(v)); });
    return apiFetch<CoverageResponse>(`/reports/coverage?${q}`);
  },

  getCoveragePdf(params: CoverageParams): Promise<Blob> {
    const q = new URLSearchParams();
    Object.entries(params).forEach(([k, v]) => { if (v !== undefined) q.set(k, String(v)); });
    q.set("format", "pdf");
    return apiFetch<Blob>(`/reports/coverage?${q}`, undefined, "blob");
  },

  getWorkload(params: WorkloadParams): Promise<WorkloadResponse> {
    const q = new URLSearchParams();
    Object.entries(params).forEach(([k, v]) => { if (v !== undefined) q.set(k, String(v)); });
    return apiFetch<WorkloadResponse>(`/reports/workload?${q}`);
  },

  getWorkloadPdf(params: WorkloadParams): Promise<Blob> {
    const q = new URLSearchParams();
    Object.entries(params).forEach(([k, v]) => { if (v !== undefined) q.set(k, String(v)); });
    q.set("format", "pdf");
    return apiFetch<Blob>(`/reports/workload?${q}`, undefined, "blob");
  },

  getDoctorDossier(doctorId: string, dateFrom?: string, dateTo?: string): Promise<DossierResponse> {
    const q = new URLSearchParams();
    if (dateFrom) q.set("date_from", dateFrom);
    if (dateTo) q.set("date_to", dateTo);
    return apiFetch<DossierResponse>(`/reports/doctor-dossier/${doctorId}?${q}`);
  },

  getDoctorDossierPdf(doctorId: string, dateFrom?: string, dateTo?: string): Promise<Blob> {
    const q = new URLSearchParams();
    if (dateFrom) q.set("date_from", dateFrom);
    if (dateTo) q.set("date_to", dateTo);
    q.set("format", "pdf");
    return apiFetch<Blob>(`/reports/doctor-dossier/${doctorId}?${q}`, undefined, "blob");
  },
};

File: backend/app/schemas/reports.py
from datetime import date
from pydantic import BaseModel


class CoverageFilterParams(BaseModel):
    year_start: int
    month_start: int
    year_end: int
    month_end: int
    area: str | None = None
    rank_id: str | None = None
    sex: str | None = None
    department_id: str | None = None


class CoverageByArea(BaseModel):
    area_id: str
    area_name: str
    days_covered: int
    days_uncovered: int
    coverage_pct: float
    gaps: list[dict]  # [{date: "2026-01-03", day_name: "Lunes"}]


class CoverageResponse(BaseModel):
    period_label: str
    overall_coverage_pct: float
    total_gaps: int
    most_critical_area: str | None
    weakest_day: str | None
    by_area: list[CoverageByArea]


class WorkloadFilterParams(BaseModel):
    year: int
    month: int
    area: str | None = None
    rank_id: str | None = None
    sex: str | None = None
    department_id: str | None = None
    group_by: str = "none"  # area|rank|department|none
    order_by: str = "total_desc"  # total_desc|alpha|rank


class DoctorWorkloadEntry(BaseModel):
    doctor_id: str
    name: str
    rank: str | None
    sex: str | None
    department: str | None
    emergencia: int = 0
    pista: int = 0
    disponible: int = 0
    total: int = 0
    details: list[dict] | None = None


class WorkloadResponse(BaseModel):
    period_label: str
    total_services: int
    active_doctors: int
    avg_per_doctor: float
    most_load: dict | None  # {name, total}
    least_load: dict | None  # {name, total}
    entries: list[DoctorWorkloadEntry]


class DossierResponse(BaseModel):
    doctor_id: str
    name: str
    rank: str | None
    sex: str | None
    department: str | None
    areas: list[str]
    period_label: str
    total_services: int
    services_by_area: dict[str, int]
    avg_weekly: float
    services: list[dict]
    missions: list[dict]
    restrictions: list[dict]
    availability: list[str]
